Extend tests for capacity policies

simkam opened the pull request:
* added FIFO behavior check to MinPoolSize, Size and  Watermark decrementer test cases
* added PoolByCri tests for MinPoolSize, Size, TimedOut, TimedOutFIFO, and Watermark decrementer

simkam commented:
@jesperpedersen can you take a look? I'll squash commits if you're fine with tests.

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -23,6 +23,8 @@
 import org.jboss.jca.core.api.connectionmanager.pool.PoolStatistics;
 import org.jboss.jca.core.connectionmanager.pool.mcp.ManagedConnectionPool;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnection;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnectionImpl;
```
No SimpleConnectionImpl - required methods needs to be exposed in SimpleConnection

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      SimpleConnection[] simpleConnections = fillPoolToSize(5);
+
+      // save string representation of ManagedConnection
+      String[] mcHashes = new String[5];
```
These are toString's - not hashes

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
Nope

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashes[i] = sci.getManagedConnection().toString();
```
An explicit toString() contract is needed then - e.g. toString() override. Or choose a better contract

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
       checkStatistics(ps, 5, 0, 2, 3);
 
+      simpleConnections = fillPoolToSize(5);
+      String[] mcHashesAfterIdleConnectionCleanup = new String[5];
```
Same deal

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
Ditto

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashesAfterIdleConnectionCleanup[i] = sci.getManagedConnection().toString();
```
Ditto

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      // check FIFO, min-pool-size=2 so last two connections in pool must remain
+      assertEquals(mcHashes[3], mcHashesAfterIdleConnectionCleanup[0]);
+      assertEquals(mcHashes[4], mcHashesAfterIdleConnectionCleanup[1]);
+
```
Add the other asserts as well

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -23,6 +23,8 @@
 import org.jboss.jca.core.api.connectionmanager.pool.PoolStatistics;
 import org.jboss.jca.core.connectionmanager.pool.mcp.ManagedConnectionPool;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnection;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnectionImpl;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -61,7 +60,18 @@ public void checkPool() throws Exception
+      SimpleConnection[] simpleConnections = fillPoolToSize(5);
+
+      // save string representation of ManagedConnection
+      String[] mcHashes = new String[5];
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -61,7 +60,18 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -61,7 +60,18 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashes[i] = sci.getManagedConnection().toString();
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -71,6 +81,19 @@ public void checkPool() throws Exception
       checkStatistics(ps, 5, 0, 3, 2);
 
+      simpleConnections = fillPoolToSize(5);
+      String[] mcHashesAfterIdleConnectionCleanup = new String[5];
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -71,6 +81,19 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -71,6 +81,19 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashesAfterIdleConnectionCleanup[i] = sci.getManagedConnection().toString();
```
As adove

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -71,6 +81,19 @@ public void checkPool() throws Exception
+      assertEquals(mcHashes[2], mcHashesAfterIdleConnectionCleanup[0]);
+      assertEquals(mcHashes[3], mcHashesAfterIdleConnectionCleanup[1]);
+      assertEquals(mcHashes[4], mcHashesAfterIdleConnectionCleanup[2]);
    }
```
Add the other asserts as well

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -23,6 +23,8 @@
 import org.jboss.jca.core.api.connectionmanager.pool.PoolStatistics;
 import org.jboss.jca.core.connectionmanager.pool.mcp.ManagedConnectionPool;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnection;
+import org.jboss.jca.core.connectionmanager.rar.SimpleConnectionImpl;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      SimpleConnection[] simpleConnections = fillPoolToSize(5);
+
+      // save string representation of ManagedConnection
+      String[] mcHashes = new String[5];
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -62,7 +61,18 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashes[i] = sci.getManagedConnection().toString();
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
       checkStatistics(ps, 5, 0, 2, 3);
 
+      simpleConnections = fillPoolToSize(5);
+      String[] mcHashesAfterIdleConnectionCleanup = new String[5];
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      for (int i = 0; i < 5; i++)
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      {
+         SimpleConnection sc = simpleConnections[i];
+         SimpleConnectionImpl sci = (SimpleConnectionImpl) sc;
+         mcHashesAfterIdleConnectionCleanup[i] = sci.getManagedConnection().toString();
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/OnePoolNoTxDecrementCapacityWatermarkNullPolicyTestCase.java:
```diff
@@ -72,5 +82,18 @@ public void checkPool() throws Exception
+      // check FIFO, null watermark => min-pool-size=2 so last two connections in pool must remain
+      assertEquals(mcHashes[3], mcHashesAfterIdleConnectionCleanup[0]);
+      assertEquals(mcHashes[4], mcHashesAfterIdleConnectionCleanup[1]);
+
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/PoolByCriNoTxDecrementCapacityMinPoolSizePolicyTestCase.java:
```diff
@@ -0,0 +1,83 @@
+   @Override
+   public void checkPool() throws Exception
+   {
+      AbstractPool pool = getPool();
```
Better description of the test case, since PoolByCri doesn't support prefil from a user PoV

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/PoolByCriNoTxDecrementCapacitySizePolicyNullTestCase.java:
```diff
@@ -0,0 +1,89 @@
+   @Override
+   public void checkPool() throws Exception
+   {
+      AbstractPool pool = getPool();
```
Better description of the test case, since PoolByCri doesn't support prefil from a user PoV

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/PoolByCriNoTxDecrementCapacityTimedOutFIFOPolicyBMTestCase.java:
```diff
@@ -0,0 +1,108 @@
+   @Override
+   public void checkPool() throws Exception
+   {
+      AbstractPool pool = getPool();
```
Better description of the test case, since PoolByCri doesn't support prefil from a user PoV

Plus what the BM rules does

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/pool/PoolByCriNoTxDecrementCapacityTimedOutPolicyBMTestCase.java:
```diff
@@ -0,0 +1,108 @@
+   @Override
+   public void checkPool() throws Exception
+   {
+      AbstractPool pool = getPool();
```
As above

jesperpedersen commented on core/src/test/java/org/jboss/jca/core/connectionmanager/rar/SimpleConnectionImpl.java:
```diff
@@ -74,4 +74,13 @@ public boolean isDetached()
    {
       return mc.isDestroyed();
    }
+
```
Find a better contract than mc.toString() - f.ex. expose a unique string

jesperpedersen commented:
A lot of white space changes too...

Feel free to squash

jesperpedersen commented:
Missing sc.close()'s too

simkam commented:
where? All connections are created and closed in PoolTestCaseAbstract#fillPoolToSize

jesperpedersen commented:
That was well hidden, it should `String[] ids = fillToPoolSize(int)` then. No access after the connection is closed

simkam commented:
* getManagedConnectionId method exposed in SimpleConnection
* toString replaced with unique id
* fillToPoolSize returns ids of ManagedConnections directly
* added missing asserts to cover entire pool
* added test description and BM rules description

jesperpedersen commented:
Rebased, and merged
